[Bank] Allow Bot Owner/Guild Owners to remove invalid users from the bank

Drapersniper opened the pull request:
### Type

- [ ] Bugfix
- [ ] Enhancement
- [x] New feature

### Description of the changes
Sinbad Please dont kill me for using a protected member of `Config` 

Add 2 new commands `removedead` which  clears all users who are in the database  but are no longer in guild

and `memberdelete` which complete delete a users bank account

Both can only be run by guild owners

Drapersniper committed:
Add command to remove dead members from bank

Drapersniper committed:
Add a global check

mikeshardmind commented:
I'm not sure this is a great idea to add prior to proper bulk interfaces in config.

If we *do* decide to add this as is, it will at minimum need a `# FIXME: use-config-bulk-update` (or some similar to ensure someone goes back for it later). I'd recommend an additional comment warning against blindly copying the method which this is done with as well to prevent people from assuming this is safe in all cases (it very much isn't).

Have not tested this, though the logic itself should be fine from a brief glance.

Drapersniper committed:
Added a FIXME so `bank_local_clean` is updated once bulk-update is implemented
Added a brief warning to warn devs not to use the `_get_base_group` as it can mess up their config files
Removed a redundant existence check

Flame442 commented on redbot/cogs/economy/economy.py:
```diff
@@ -247,6 +247,45 @@ def __init__(self, bot: Red):
+    @commands.guild_only()
+    @checks.guildowner()
+    async def removedead(self, ctx, confirmation: bool = False):
+        """Delete bank accounts for users no longer in the server"""
```
This is missing a period.

Drapersniper replied:
Added Period

Flame442 commented on redbot/cogs/economy/economy.py:
```diff
@@ -247,6 +247,45 @@ def __init__(self, bot: Red):
+    async def removedead(self, ctx, confirmation: bool = False):
+        """Delete bank accounts for users no longer in the server"""
+        if await bank.is_global():
+            return await ctx.send(_("This command will not work in a Global bank"))
```
Should be changed to something like "This command cannot be used with a global bank." so that it is in the present tense and is more in line with other strings in core.

Also, this is missing a period.

Drapersniper replied:
Updated string to bring it in line with other strings in core and added period

Flame442 commented on redbot/cogs/economy/economy.py:
```diff
@@ -247,6 +247,45 @@ def __init__(self, bot: Red):
+        if confirmation is False:
+            await ctx.send(
+                _(
+                    "This will delete all bank accounts for user no longer in this server."
```
"... for user**s** no..."

Flame442 commented on redbot/cogs/economy/economy.py:
```diff
@@ -247,6 +247,45 @@ def __init__(self, bot: Red):
+            )
+        else:
+            await bank.bank_local_clean(guild=ctx.guild)
+            await ctx.send(_("Bank accounts for user no longer in this server have been deleted."))
```
"...for user**s** no..."

Flame442 commented on redbot/cogs/economy/economy.py:
```diff
@@ -247,6 +247,45 @@ def __init__(self, bot: Red):
+    @commands.guild_only()
+    @checks.guildowner()
+    async def memberdelete(self, ctx, member: discord.Member, confirmation: bool = False):
+        """Delete bank for specified member."""
```
Should probably be changed to """Delete the bank account of a specified member."""

Drapersniper replied:
Updated String

Flame442 commented on redbot/cogs/economy/economy.py:
```diff
@@ -247,6 +247,45 @@ def __init__(self, bot: Red):
+        if confirmation is False:
+            await ctx.send(
+                _(
+                    "This will delete {member} bank account."
```
Should probably be "...delete {member}'s bank..."

Flame442 commented on redbot/cogs/economy/economy.py:
```diff
@@ -247,6 +247,45 @@ def __init__(self, bot: Red):
+                _(
+                    "This will delete {member} bank account."
+                    "\nIf you're sure, type "
+                    "`{prefix}bank memberdelete '{member}' yes`"
```
Why is `{member}` wrapped in single quotes? (`'`)

Drapersniper replied:
This was to provide the users with a simply Copy/Paste line for the command as some users may have spaces in their names and without the quote the command would fail

Flame442 replied:
Single quotes do not work for this like double quotes do. Single quotes are taken literally. You could use the user ID so that you do not need to worry about spaces, or swap the single quotes to double quotes and escape them.

Flame442 commented on redbot/cogs/economy/economy.py:
```diff
@@ -247,6 +247,45 @@ def __init__(self, bot: Red):
+            )
+        else:
+            await bank.bank_local_clean(guild=ctx.guild, member=member)
+            await ctx.send(_("Bank account for {member} has been deleted.").format(member=member))
```
Should probably be changed to "**The** bank account..."

Flame442 commented on redbot/core/bank.py:
```diff
@@ -332,6 +332,26 @@ def _invalid_amount(amount: int) -> bool:
 
 
+async def bank_local_clean(guild: discord.Guild, member: discord.member = None):
+    """Remove data for members who are no longer in the server"""
```
You need to document what this does better. This doc does not even state that it can also be used to delete a specific member's bank account.

Flame442 commented on redbot/core/bank.py:
```diff
@@ -332,6 +332,26 @@ def _invalid_amount(amount: int) -> bool:
+    """Remove data for members who are no longer in the server"""
+    # FIXME: use-config-bulk-update
+    # We do not recommend that devs use this protected method as it is not safe to be used in
+    # all cased and this can end up corrupting your config file
```
This is a run on and has poor grammar.

Drapersniper replied:
Think this is improved now let me know if you disagree and suggestions around it

Flame442 commented on redbot/core/bank.py:
```diff
@@ -332,6 +332,26 @@ def _invalid_amount(amount: int) -> bool:
+        accounts = await group.all()
+        tmp = accounts.copy()
+    async with group.all() as member_data:
+        if member is not None:
```
You have this backwards, the first have is `if member is not None`, so there is a `member`. However, you clean the entire server here. In the `else` where there is no `member`, you attempt to clean a specific member.

Flame442 commented on redbot/core/bank.py:
```diff
@@ -332,6 +332,26 @@ def _invalid_amount(amount: int) -> bool:
+                if not guild.get_member(acc):
+                    del member_data[acc]
+        else:
+            member_id = f"{member.id}"
```
Are you converting `member.id` to a string by using an f string here?

Flame442 commented on redbot/cogs/economy/economy.py:
```diff
@@ -247,6 +247,45 @@ def __init__(self, bot: Red):
+    @_bank.command()
+    @commands.guild_only()
+    @checks.guildowner()
+    async def removedead(self, ctx, confirmation: bool = False):
```
I don't love this command name as it isn't very clear what this will do (it is unclear what "dead" means).

Drapersniper replied:
Read below: https://github.com/Cog-Creators/Red-DiscordBot/pull/2845#discussion_r301943785

Flame442 commented on redbot/cogs/economy/economy.py:
```diff
@@ -247,6 +247,45 @@ def __init__(self, bot: Red):
+    async def memberdelete(self, ctx, member: discord.Member, confirmation: bool = False):
+        """Delete bank for specified member."""
+        if await bank.is_global():
+            return await ctx.send(_("This command will not work in a Global bank"))
```
See above.

Drapersniper replied:
I've created a `cleanup` group with 2 sub-commands "member" and "server" is this approach better?

palmtree5 replied:
ehh, please don't name it `cleanup` as that command is the entire basis of a core cog

EDIT: I probably should've looked at the code first. It wasn't clear to me without doing so that this `cleanup` is a subgroup of `[p]bank` in which case my comment is irrelevant

Drapersniper replied:
I've updated the group name to `prune` and split the commands into 3

[p]bank prune local  : Can be run by Guild owners only
[p]bank prune global : Can be run by Bot Owner only
[p]bank prune user   : Can be run by Admins, Guild owners and Bot Owner

Flame442 reviewed (changes requested):
I still think this needs some more discussion, however I have reviewed the code as-is so it can be improved.

Additional comments that don't apply to a particular line:
-All of the places you use `member=member`, you might want to use `member=member.display_name` instead so that it will use nicknames.
-It would be nice if this supported global banks and had an extra check that the invoker is a bot owner when the bank is global.

Drapersniper committed:
Updated  commit to reflect changes requested in review

Drapersniper committed:
Updated  commit to reflect changes requested in review

Drapersniper committed:
Merge remote-tracking branch 'origin/removedead' into removedead

# Conflicts:
#	redbot/cogs/economy/economy.py

Drapersniper committed:
:facepalm:

Drapersniper committed:
Return command to run with user id so we don't worry about safeguarding the command agaisn't invalid formats

Drapersniper committed:
Braaaainnn

Removed aliases that used old naming scheme

Drapersniper commented:
> -It would be nice if this supported global banks and had an extra check that the invoker is a bot owner when the bank is global.

I agree but how would this work? Since Global bank means the balance works across servers.

Or do you mean that the `cleanup member` command should be  be available to the bot owner?

Flame442 commented:
If the bank is global, accounts will be tied to Users (rather than Members). Bot owners should be able to delete a user's account when the bank is global and possibly clear all users that no longer share a guild with the bot.

Drapersniper committed:
TL:DR  Added global bank support, and rework permissions

Renamed `bank_local_clean` to `bank_prune`
Added support for global banks to `bank_prune`
`bank_prune` will now raise `BankPruneError` if trying to prune a local bank and `guild` is not supplied

Renamed `cleanup` subgroup to `prune`
`prune` subgroup will have 3 commands:
  `user`   :  Deletes the bank account for the specified member : Accepts `Union[discord.Member, discord.User, int]`
  `global` :  Prune global bank accounts for all users who no longer share a server with the bot
  `local`  :  Prune local bank accounts for all users who are no longer in the guild

Changed check for `prune` subgroup to be `@check_global_setting_admin()`
[p]bank prune local  : Can be run by Guild owners only
[p]bank prune global : Can be run by Bot Owner only
[p]bank prune user   : Can be run by Admins, Guild owners and Bot Owner

Drapersniper commented:
Global support has been added, also reworked perms and names:

[p]bank prune local  : Can be run by Guild owners only
[p]bank prune global : Can be run by Bot Owner only
[p]bank prune user   : Can be run by Admins, Guild owners and Bot Owner

Drapersniper committed:
Yikes ... Updated kwarg name

Drapersniper committed:
Fixed unexpected unindent: docstring of redbot.core.bank.bank_prune:14:Field list ends without a blank line

Drapersniper committed:
...

Drapersniper committed:
3rd time lucky?

Drapersniper committed:
4th time lucky?

Drapersniper committed:
Fix Docstring

Flame442 commented on redbot/cogs/economy/economy.py:
```diff
@@ -247,6 +247,82 @@ def __init__(self, bot: Red):
+
+    @_prune.command()
+    async def user(
+        self, ctx, user: Union[discord.Member, discord.User, int], confirmation: bool = False
```
I started some discussion in #coding about this, I'm not sure exactly what you should use for `user`, but I don't think this is it.

Flame442 commented on redbot/cogs/economy/economy.py:
```diff
@@ -247,6 +247,82 @@ def __init__(self, bot: Red):
+            )
+
+    @_prune.command()
+    async def user(
```
This should be able to delete a user's account in a global bank, but **limited to the bot owner**.

Drapersniper replied:
This should already do this for global banks, due to the `check_global_setting_admin` check i added to the group

Flame442 commented on redbot/core/bank.py:
```diff
@@ -332,6 +335,64 @@ def _invalid_amount(amount: int) -> bool:
 
 
+async def bank_prune(
+    bot: Red, guild: Optional[discord.Guild] = None, user_id: Optional[int] = None
```
If you want to use these as keyword params, make them keyword params instead of `Optional`.

Flame442 commented on redbot/core/bank.py:
```diff
@@ -332,6 +335,64 @@ def _invalid_amount(amount: int) -> bool:
+    bot : Red
+        The bot.
+    guild : discord.Guild
+        The guild to prune this is required if bank is set to local.
```
"The guild to prune. **T**his is required if **the** bank is set to local."

Flame442 commented on redbot/core/bank.py:
```diff
@@ -332,6 +335,64 @@ def _invalid_amount(amount: int) -> bool:
+    guild : discord.Guild
+        The guild to prune this is required if bank is set to local.
+    user_id : int
+        The id of the user who's account will be pruned.
```
"who's" should be "whose".

Flame442 commented on redbot/core/bank.py:
```diff
@@ -332,6 +335,64 @@ def _invalid_amount(amount: int) -> bool:
+
+    Notes
+    -----
+        :code:`Config._get_base_group()` should not be used to get config groups as
```
I understand that this is used in the code, but putting this here with no explanation as to why you are saying this makes no sense to someone only reading the docs (not the code).

Drapersniper replied:
I've moved this to the definition of `_get_base_group` so it will show in it's doc rather than here think that will make more sense

Flame442 commented on redbot/core/bank.py:
```diff
@@ -332,6 +335,64 @@ def _invalid_amount(amount: int) -> bool:
+    Notes
+    -----
+        :code:`Config._get_base_group()` should not be used to get config groups as
+        this is not a safe operation using this could end up corrupting your config file.
```
"...a safe operation. **U**sing this could..."

Flame442 commented on redbot/core/bank.py:
```diff
@@ -332,6 +335,64 @@ def _invalid_amount(amount: int) -> bool:
+    global_bank = await is_global()
+
+    if global_bank:
+        _guilds = [g for g in bot.guilds if g.large and not (g.chunked or g.unavailable)]
```
Not sure why you only want large guilds / what this is supposed to do.

Flame442 commented on redbot/core/errors.py:
```diff
@@ -49,3 +49,9 @@ def __str__(self) -> str:
+
+class BankPruneError(BankError):
+    """Raised when trying to prune a local bank and no server is specified."""
+
```
Does Black yell without a new line here?

Drapersniper replied:
Yeah it requires 2 blank lines between classes

Flame442 reviewed (changes requested):
After deleting my bank account with `[p]bank prune user Flame yes`, `[p]bank balance` returns this error (I'm not cropping it because I don't know how much is useful and I'm too tired to check):
```py
Exception in command 'bank balance'
Traceback (most recent call last):
  File "c:\Users\Flame\Desktop\Random\TestingRedbot\reddev\VENV\lib\site-packages\discord\ext\commands\core.py", line 63, in wrapped
    ret = await coro(*args, **kwargs)
  File "c:\Users\Flame\Desktop\Random\TestingRedbot\reddev\VENV\lib\site-packages\redbot\cogs\economy\economy.py", line 157, in balance
    bal = await bank.get_balance(user)
  File "c:\Users\Flame\Desktop\Random\TestingRedbot\reddev\VENV\lib\site-packages\redbot\core\bank.py", line 139, in get_balance
    acc = await get_account(member)
  File "c:\Users\Flame\Desktop\Random\TestingRedbot\reddev\VENV\lib\site-packages\redbot\core\bank.py", line 492, in get_account
    all_accounts = await _conf.all_users()
  File "c:\Users\Flame\Desktop\Random\TestingRedbot\reddev\VENV\lib\site-packages\redbot\core\config.py", line 1057, in all_users
    return await self._all_from_scope(self.USER)
  File "c:\Users\Flame\Desktop\Random\TestingRedbot\reddev\VENV\lib\site-packages\redbot\core\config.py", line 987, in _all_from_scope
    ret[int(k)] = data
ValueError: invalid literal for int() with base 10: 'name'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "c:\Users\Flame\Desktop\Random\TestingRedbot\reddev\VENV\lib\site-packages\discord\ext\commands\bot.py", line 860, in invoke
    await ctx.command.invoke(ctx)
  File "c:\Users\Flame\Desktop\Random\TestingRedbot\reddev\VENV\lib\site-packages\redbot\core\commands\commands.py", line 570, in invoke
    await super().invoke(ctx)
  File "c:\Users\Flame\Desktop\Random\TestingRedbot\reddev\VENV\lib\site-packages\discord\ext\commands\core.py", line 1127, in invoke
    await ctx.invoked_subcommand.invoke(ctx)
  File "c:\Users\Flame\Desktop\Random\TestingRedbot\reddev\VENV\lib\site-packages\discord\ext\commands\core.py", line 698, in invoke
    await injected(*ctx.args, **ctx.kwargs)
  File "c:\Users\Flame\Desktop\Random\TestingRedbot\reddev\VENV\lib\site-packages\discord\ext\commands\core.py", line 72, in wrapped
    raise CommandInvokeError(exc) from exc
discord.ext.commands.errors.CommandInvokeError: Command raised an exception: ValueError: invalid literal for int() with base 10: 'name'
```
This same error appears for other commands that interact with the bank, such as `[p]payday`. I can't really test this effectively with this happening.

Drapersniper committed:
Initial commit to address review by Flame

Drapersniper committed:
Updated code to reflect Flame's comments

Drapersniper committed:
Skip pruning of unavailable guilds

Drapersniper committed:
Fixed typo in string

Drapersniper committed:
*sigh* black is the bane of my existence

Flame442 commented on redbot/cogs/economy/economy.py:
```diff
@@ -247,6 +248,80 @@ def __init__(self, bot: Red):
+            )
+        else:
+            await bank.bank_prune(self.bot, guild=ctx.guild)
+            await ctx.send(_("Bank accounts for users no longer in this serve have been deleted."))
```
"serve" -> "server"

Flame442 commented on redbot/cogs/economy/economy.py:
```diff
@@ -247,6 +248,80 @@ def __init__(self, bot: Red):
+    @_prune.command(name="global")
+    @checks.is_owner()
+    async def _global(self, ctx, confirmation: bool = False):
+        """Prune bank accounts for users who no longer share servers with the bot."""
```
Instead of "share servers", maybe "share a server".

Flame442 commented on redbot/cogs/economy/economy.py:
```diff
@@ -247,6 +248,80 @@ def __init__(self, bot: Red):
+            await ctx.send(
+                _(
+                    "Bank accounts for users who "
+                    "no longer share servers with the bot have been pruned."
```
Same comment, maybe "share a server".

Flame442 commented on redbot/cogs/economy/economy.py:
```diff
@@ -247,6 +248,80 @@ def __init__(self, bot: Red):
+            )
+
+    @_prune.command()
+    async def user(self, ctx, user: Union[discord.Member, RawUserIds], confirmation: bool = False):
```
The var `user` is used twice (as discussed in #testing)

Flame442 commented on redbot/core/bank.py:
```diff
@@ -332,6 +335,59 @@ def _invalid_amount(amount: int) -> bool:
+    bot : Red
+        The bot.
+    guild : discord.Guild
+        The guild to prune this is required if the bank is set to local.
```
"prune. **T**his"

Flame442 reviewed (changes requested):
Looking a lot better.

Jackenmen commented on redbot/cogs/economy/economy.py:
```diff
@@ -247,6 +248,80 @@ def __init__(self, bot: Red):
+                ).format(prefix=ctx.prefix, id=uid, name=name)
+            )
+        else:
+            await bank.bank_prune(self.bot, guild=ctx.guild, user_id=uid)
```
`ctx.guild` might be `None` if command is used in DMs.

Drapersniper replied:
Will now return an error when trying to run the command via DM if the user has a local bank

Drapersniper committed:
addressed Flames commends
Fixed [p]bank prune user, When run via DM it will now return an error message to the user (Thanks jack1142)

Drapersniper committed:
Time to get some sleep

Drapersniper committed:
'DM' > 'DMs' in string

Flame442 reviewed (dismissed):
All of the things I wanted to see fixed have been fixed. Thanks for working with me.

Drapersniper committed:
Merge branch 'V3/develop' into removedead

Drapersniper committed:
Merge branch 'V3/develop' into removedead

Drapersniper committed:
Merge branch 'V3/develop' into removedead

Drapersniper committed:
Merge branch 'V3/develop' into removedead

Drapersniper committed:
Merge branch 'removedead' of https://github.com/Drapersniper/Red-DiscordBot into removedead

Signed-off-by: Draper <guyreis96@gmail.com>

# Conflicts:
#	redbot/cogs/economy/economy.py
#	redbot/core/bank.py
#	redbot/core/config.py

Drapersniper committed:
Add towncrier entries

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Merge branch 'V3/develop' into removedead

Drapersniper committed:
Merge branch 'V3/develop' into removedead

Drapersniper committed:
Merge branch 'V3/develop' into removedead

Flame442 commented on changelog.d/2845.feature.rst:
```diff
@@ -0,0 +1,5 @@
+Added New commands to Economy
+
+- ``[p]bank prune user`` - This will delete a users account
```
You should specify that it deletes a user's *bank* account.

Flame442 commented on changelog.d/2845.feature.rst:
```diff
@@ -0,0 +1,5 @@
+Added New commands to Economy
+
+- ``[p]bank prune user`` - This will delete a users account
+- ``[p]bank prune local`` - This will prune a local bank from users who are no longer in the server
```
This should specify that this will prune the bank *of accounts* from users no longer in the server.

Flame442 commented on changelog.d/2845.misc.rst:
```diff
@@ -0,0 +1,3 @@
+Added :func:`bank_prune` to :module:`redbot.core.bank`
```
As this function will be documented, this should not be a misc changelog. Only undocumented and internal changes should be labeled as misc.

Flame442 commented on changelog.d/2845.feature.rst:
```diff
@@ -0,0 +1,5 @@
+
+- ``[p]bank prune user`` - This will delete a users account
+- ``[p]bank prune local`` - This will prune a local bank from users who are no longer in the server
+- ``[p]bank prune global`` - This will prune the global bank from users who do not share any servers with the bot
```
Just like above, this will prune the global bank *of accounts* from users...

Flame442 commented on changelog.d/2845.misc.rst:
```diff
@@ -0,0 +1,3 @@
+Added :func:`bank_prune` to :module:`redbot.core.bank`
+
+- :func:`bank_prune` can be used to delete a users bank account or remove all users who are not in the guild (bank is local) or visible to the bot (bank is global)
```
This sentence has too many `or`s that it isn't clear what you are trying to say. The situation is `A or (B or C)`, however that isn't very clear from the wording.

Flame442 reviewed (changes requested):
New changes requested for the changelog.

Drapersniper committed:
Update to reflect Flame's review

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>
